cellarium/ml/layers/attention.py

```diff
@@ -0,0 +1,164 @@
+
+import torch
+from torch import nn
+from torch.nn.attention import SDPBackend, sdpa_kernel
+
+from cellarium.ml.utilities.layers import create_initializer
+
+try:
+    # use_cs returns True if the active device is a CSX device.
+    from cerebras.pytorch.backend import use_cs
+except ImportError:
+
+    def use_cs() -> bool:
+        return False
+
+
+class MultiHeadAttention(nn.Module):
+    """
+    Multi-head attention.
+
+    Args:
+        d_model:
+            Dimensionality of the embeddings and hidden states.
+        use_bias:
+            Whether to use bias in the linear transformations.
+        n_heads:
+            Number of attention heads.
+        dropout_p:
+            Dropout probability.
+        attention_logits_scale:
+            Multiplier for the attention scores.
+        attention_backend:
+            Backend for the attention computation.
+        attention_softmax_fp32:
+            Whether to use float32 for softmax computation.
+        Wqkv_initializer:
+            Initializer for the query, key, and value linear transformations.
+        Wo_initializer:
+            Initializer for the output linear transformation.
+    """
+
+    backend_map = {
+        "math": SDPBackend.MATH,
+        "flash": SDPBackend.FLASH_ATTENTION,
+        "mem_efficient": SDPBackend.EFFICIENT_ATTENTION,
+    }
+
+    def __init__(
+        self,
+        d_model: int,
+        use_bias: bool,
+        n_heads: int,
+        dropout_p: float,
+        attention_logits_scale: float,
+        attention_backend: Literal["math", "flash", "mem_efficient", "torch"],
+        attention_softmax_fp32: bool,
+        Wqkv_initializer: dict[str, Any],
+        Wo_initializer: dict[str, Any],
+    ) -> None:
+        super().__init__()
+        self.Wq = nn.Linear(d_model, d_model, bias=use_bias)
+        self.Wk = nn.Linear(d_model, d_model, bias=use_bias)
+        self.Wv = nn.Linear(d_model, d_model, bias=use_bias)
+        self.Wo = nn.Linear(d_model, d_model, bias=use_bias)
+        self.n_heads = n_heads
+        self.dropout_p = dropout_p
+        self.attention_logits_scale = attention_logits_scale
+        self.attention_backend = attention_backend
+        self.attention_softmax_fp32 = attention_softmax_fp32
+        self.Wqkv_initializer = Wqkv_initializer
+        self.Wo_initializer = Wo_initializer
+
+        self._reset_parameters()
+
+    def _reset_parameters(self) -> None:
+        for module in [self.Wq, self.Wk, self.Wv]:
+            create_initializer(self.Wqkv_initializer)(module.weight)
+            if module.bias is not None:
+                nn.init.zeros_(module.bias)
+
+        create_initializer(self.Wo_initializer)(self.Wo.weight)
+        if self.Wo.bias is not None:
+            nn.init.zeros_(self.Wo.bias)
+
+    @staticmethod
+    def split_heads(X_nqd: torch.Tensor, n_heads: int) -> torch.Tensor:
+        """Transposition for parallel computation of multiple attention heads."""
```

Comment:
Add `assert X_nqd.shape[-1] % n_heads == 0`, here or better yet, in the initializer of the `TransformerBlock`.